eip_operator/src/egress.rs

```diff
@@ -0,0 +1,152 @@
+    }
+
+    // Build up a list of EIP resourceId's to check EIP attachment against node names.
+    let eip_resource_ids: BTreeSet<String> = eip_api
+        .list(&ListParams::default())
+        .await?
+        .items
+        .into_iter()
+        .filter_map(|eip| eip.status.and_then(|s| s.resource_id))
+        .collect();
+
+    // Build up a map of egress node names along with their
+    // ready status and whether or not they have an EIP attached.
+    let mut node_map: BTreeMap<String, (String, bool)> = BTreeMap::new();
+    for node in egress_nodes {
+        if let Some(ref node_name) = node.metadata.name {
+            let ready_status = node
+                .status
+                .as_ref()
+                .and_then(|status| status.conditions.as_ref())
+                .and_then(|conditions| conditions.iter().find(|c| c.type_ == "Ready"))
+                .map(|condition| condition.status.clone())
+                .ok_or(Error::MissingNodeReadyCondition)?;
+            let eip_attached_status = eip_resource_ids.contains(node_name.as_str());
+            node_map.insert(node_name.clone(), (ready_status, eip_attached_status));
+        }
+    }
+
+    let egress_nodes_ready_with_eip: BTreeMap<String, (String, bool)> = node_map
+        .iter()
+        .filter(|(_, &(ref ready_status, eip_attached_status))| {
+            ready_status == "True" && eip_attached_status
+        })
+        .map(|(node_name, &(ref ready_status, eip_attached_status))| {
+            (
+                node_name.clone(),
+                (ready_status.clone(), eip_attached_status),
+            )
+        })
+        .collect();
+
+    // Wait to label egress nodes until an EIP is attached.
+    // Setting the ready status label to "ready" routes egress traffic through the node
+    // Wait to label egress nodes until they are ready and have an EIP attached.
+    if egress_nodes_ready_with_eip.is_empty() {
+        info!(
+            "No egress nodes found with a ready status and attached EIP. Skipping egress labeling."
+        );
+        return Ok(());
+    }
+
+    // At least one egress node should be ready with an EIP attached in this current implementation.
+    // This allows the ready status label to be set and traffic to be routed through the node.
+    assert!(egress_nodes_ready_with_eip.len() == 1);
+    if let Some((first_key, _)) = egress_nodes_ready_with_eip.first_key_value() {
+        let node_name = first_key.clone();
+        add_gateway_status_label(node_api, node_name.as_str(), "true").await?;
+    }
+
+    // We immediately disassociate EIPs from egress nodes that are in an
+    // unresponsive state in the eip node controller.
+    // This allows the EIP to be re-associated with a new healthy node.
+    // However, unresponsive nodes may still exist with a ready-status egress label
+    // set to "true". This allows the old node to still serve traffic if possible until a
+    // new node is ready to take over.
```

Comment:
Discussed over slack and covered by the egress labeling and node controller refactors. We agreed to set the status of the EIP to detached to free the claim. Egress will continue to attempt to go out the unresponsive node on the EIP as it will still be attached to the instance.